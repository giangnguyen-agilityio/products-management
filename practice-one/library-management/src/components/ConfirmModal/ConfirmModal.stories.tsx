import { type StoryFn, type Meta } from "@storybook/react";
import ConfirmModal from "./index";

export default {
  title: "Example/ConfirmModal",
  component: ConfirmModal,
  argTypes: {
    onConfirm: { action: "Confirm button clicked" },
    onCancel: { action: "Cancel button clicked" },
  },
} as Meta<typeof ConfirmModal>;

const Template: StoryFn<typeof ConfirmModal> = (args) => (
  <ConfirmModal {...args} />
);

export const Default = Template.bind({});
Default.args = {
  text: "book",
};
