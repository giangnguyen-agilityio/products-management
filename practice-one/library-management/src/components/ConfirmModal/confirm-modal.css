.confirm-modal-content {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
}

.confirm-modal-image {
  width: 250px;
}

.confirm-modal-title {
  font-family: var(--secondary-font);
  font-size: 20px;
  line-height: 1.2;
  margin: 10px 0 30px;
  color: var(--secondary-color);
}

.highlight-text {
  color: var(--danger-color);
}

.confirm-modal-control {
  display: flex;
  gap: 60px;
}

.confirm-modal-button,
.cancel-modal-button {
  width: 100px;
  transition: all 0.3s ease-in-out;
  opacity: 0.8;
  border: 1px solid transparent;
}

.cancel-modal-button {
  color: var(--secondary-color);
  background-color: unset;
}

.cancel-modal-button:hover {
  opacity: unset;
  border-color: var(--secondary-color);
}

.confirm-modal-button {
  background-color: var(--danger-color);
  color: var(--secondary-text-color);
}

.confirm-modal-button:hover {
  opacity: unset;
}
