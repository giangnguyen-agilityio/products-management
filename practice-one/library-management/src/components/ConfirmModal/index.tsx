import React, { useContext, useState } from 'react'

// Importing the Typography, and Button components
import Button from '../Button'
import Typography from '../Typography'

// Importing the BookContext
import BookContext from '../../store/BookContext'

// Import the Delete image
import deleteImage from '../../assets/images/delete-icon.gif'

// Importing the API method
import { deleteBookAPI } from '../../services/api-actions'

// Importing the actions
import { deleteBook } from '../../store/action'

// Importing the CSS file for styling
import './confirm-modal.css'

// Define the props for the Confirm modal component
interface ConfirmModalProps {
  id: string
  text: string
  onCloseModal: (showToast?: boolean) => void
  handleToast: (message: string, status: 'success' | 'failure') => void
}

const ConfirmModal: React.FC<ConfirmModalProps> = (props) => {
  const { id, text, onCloseModal, handleToast } = props

  // Get the dispatch function from BookContext
  const { dispatch } = useContext(BookContext)

  // State to set disable button
  const [disableButton, setDisableButton] = useState(false)

  // Function to handle confirm (delete)
  const handleDelete = async (): Promise<void> => {
    try {
      setDisableButton(true) // Set disableButton state to true
      // Call deleteBookAPI with the book ID
      const result = await deleteBookAPI(id)
      if (result != null) {
        setDisableButton(false) // Set disableButton state to false
        // Dispatch the deleteBook action with the book ID
        dispatch(deleteBook(id))
        // Show a success toast message
        handleToast('Book deleted successfully', 'success')
        // Close the modal
        onCloseModal()
      }
    } catch (error) {
      // Show a failure toast message if deletion fails
      handleToast('Failed to delete book', 'failure')
      // Close the modal
      onCloseModal()
    }
  }

  // Function to handle cancel
  const handleCancel = (): void => {
    // Close the modal without showing the toast
    onCloseModal(false)
  }

  return (
    <div className="confirm-modal-content">
      <img src={deleteImage} alt="Image for delete action" className="confirm-modal-image" />
      <Typography variant="p" className="confirm-modal-title">
        Do you really want to delete this {text} ?
        <br />
        This process{' '}
        <strong className="highlight-text"> cannot be undone.</strong>
      </Typography>
      <div className="confirm-modal-control">
        <Button
          variant="primary"
          className="confirm-modal-button"
          onClick={handleDelete}
          disabled={disableButton}
        >
          Delete
        </Button>
        <Button
          variant="primary"
          className="cancel-modal-button"
          onClick={handleCancel}
        >
          Cancel
        </Button>
      </div>
    </div>
  )
}

export default ConfirmModal
