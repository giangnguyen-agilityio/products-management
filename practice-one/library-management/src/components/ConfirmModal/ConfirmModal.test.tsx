import React from 'react'
import {render, screen, fireEvent} from '@testing-library/react'
import ConfirmModal from '.'

jest.mock('@assets/images/delete-icon.gif', () => ({
  default: 'delete-icon.gif',
}))

describe('ConfirmModal', () => {
  const mockId = '123'
  const mockText = 'item'
  const mockOnCloseModal = jest.fn()
  const mockOnDelete = jest.fn()

  it('should renders the component with the correct props', () => {
    const {container} = render(
      <ConfirmModal
        id={mockId}
        text={mockText}
        onCloseModal={mockOnCloseModal}
        onDelete={mockOnDelete}
      />
    )

    // Check if the image is rendered with the correct attributes
    const deleteImage = screen.getByAltText('Image for delete action')
    expect(deleteImage).toBeInTheDocument()
    expect(deleteImage).toHaveAttribute('src', 'delete-icon.gif')
    expect(container).toMatchSnapshot()
  })

  it('should calls the onDelete function when clicking the Delete button', () => {
    render(
      <ConfirmModal
        id={mockId}
        text={mockText}
        onCloseModal={mockOnCloseModal}
        onDelete={mockOnDelete}
      />
    )

    // Simulate clicking the "Delete" button
    const deleteButton = screen.getByText('Delete')
    fireEvent.click(deleteButton)

    // Check if the onDelete function is called with the correct id
    expect(mockOnDelete).toHaveBeenCalledWith(mockId)
  })

  it('should calls the onCloseModal function when clicking the Cancel button', () => {
    render(
      <ConfirmModal
        id={mockId}
        text={mockText}
        onCloseModal={mockOnCloseModal}
        onDelete={mockOnDelete}
      />
    )

    // Simulate clicking the "Cancel" button
    const cancelButton = screen.getByText('Cancel')
    fireEvent.click(cancelButton)

    // Check if the onCloseModal function is called
    expect(mockOnCloseModal).toHaveBeenCalled()
  })
})
