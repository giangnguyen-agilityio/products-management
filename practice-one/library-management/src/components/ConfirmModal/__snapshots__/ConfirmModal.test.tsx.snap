// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal should renders the component with the correct props 1`] = `
<div>
  <div
    class="confirm-modal-content"
  >
    <img
      alt="Image for delete action"
      class="confirm-modal-image"
      src="delete-icon.gif"
    />
    <p
      class="text-p text-default confirm-modal-title"
    >
      Do you really want to delete this 
      item
       ?
      <br />
      This process
       
      <strong
        class="highlight-text"
      >
         cannot be undone.
      </strong>
    </p>
    <div
      class="confirm-modal-control"
    >
      <button
        class="btn btn-primary btn-medium confirm-modal-button"
        type="button"
      >
        Delete
      </button>
      <button
        class="btn btn-primary btn-medium cancel-modal-button"
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
